app/src/main/kotlin/compiler/analysis/TypeChecker.kt

```diff
@@ -137,7 +148,7 @@ class TypeChecker(private val nameResolution: Map<Ref<AstNode>, Ref<NamedNode>>,
 
         checkBlock(function.body)
 
-        if (function.implementation is Function.Implementation.Local && function.returnType != Type.Unit)
+        if (function.implementation is Function.Implementation.Local && function.returnType != Type.Unit && !function.isGenerator)
```

Comment:
Similar extra condition (`&& !function.isGenerator`) will be needed in `compiler.analysis.ReturnValueVariableCreator:17`